### PR TITLE
[Schema] Fix autoConsumeSchema deadLetter problem.

### PR DESCRIPTION
**Pull request by congbobo184**
## Motivation
in order to fix autoConsumeSchema deadLetter problem.
fix https://github.com/apache/pulsar/issues/9935

## implement
Add AutoProduceSchema to fix it
### Verifying this change
Add the tests for it

Does this pull request potentially affect one of the following parts:
If yes was chosen, please highlight the changes

Dependencies (does it add or upgrade a dependency): (no)
The public API: (no)
The schema: (no)
The default values of configurations: (no)
The wire protocol: (no)
The rest endpoints: (no)
The admin cli options: (no)
Anything that affects deployment: (no)

### CODE REVIEW

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoProduceSchema.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+        if (requireSchemaValidation) {
+            // verify the message can be detected by the underlying schema
+            schema.decode(bytes, schemaVersion);
+        }
+        return schema.decode(bytes, schemaVersion);
+    }
+
+    @Override
+    public SchemaInfo getSchemaInfo() {
+        return BytesSchema.of().getSchemaInfo();
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoProduceSchema.java:82]: what about this version ?
IMHO we should delegate to the wrapped schema

```
@Override
    public SchemaInfo getSchemaInfo() {
        ensureSchemaInitialized();

        return schema.getSchemaInfo();
    }
```

**Review by eolivelli — CHANGES_REQUESTED** [review]
awesome work @congbobo184 

I see that this new class is very similar to AutoProduceBytesSchema

is there a way to consolidate the logic in a common base class

**Comment by congbobo184** [comment]
@eolivelli thanks for you comment, I change the deadLetterProducer to AutoProduceBytesSchema. please review again.

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java [hunk 2]
```diff
@@ -114,6 +117,21 @@ public SchemaInfo getSchemaInfo() {
         return schema.getSchemaInfo();
     }
 
+    @Override
+    public CompletableFuture<SchemaInfo> getSchemaInfo(byte[] schemaVersion) {
+        ensureSchemaInitialized();
+        if (schema.supportSchemaVersioning()) {
+            if (schemaVersion == null) {
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java:124]: isn't is better to consistently return an error if schemaVersion is null ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java [hunk 3]
```diff
@@ -319,10 +321,17 @@ public boolean publishedEarlierThan(long timestamp) {
         }
     }
 
-    public Schema<T> getSchema() {
+    public Schema<T> getCurrentSchema() {
         return this.schema;
     }
 
+    public CompletableFuture<Schema<?>> getOriginalSchema() {
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/MessageImpl.java:328]: can you please add a javadoc that explains why method returns a CompletableFuture ?
for instance we could need to download the schema from the network.

can we name this method "fetchSchema" ?
This way the name itself reads that this is not a simple "getter" but that the operation may perform additional operations.

**Review by eolivelli — CHANGES_REQUESTED** [review]

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java (hunk 2 above)
Comment (congbobo184) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/AutoConsumeSchema.java:124]: good idea

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java [hunk 4]
```diff
@@ -88,15 +88,17 @@
 
     private interface PulsarSinkProcessor<T> {
```
Comment (eolivelli) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java:89]: why are we changing PulsarSink ?

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java [hunk 5]
```diff
@@ -425,46 +426,58 @@ public void sendAsync(Message<?> message, SendCallback callback) {
             return;
         }
 
-        if (!populateMessageSchema(msg, callback)) {
```
Comment (eolivelli) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:428]: is it necessary to change ProducerImpl ?

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java [hunk 6]
```diff
@@ -48,7 +48,7 @@
         return Optional.empty();
     }
 
-    default Schema<T> getSchema() {
+    default Schema getSchema() {
```
Comment (eolivelli) [inline pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:51]: why are we removing this template parameter ?
it is wrong, if the Record is of type T, than the Schema is for type T

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java [hunk 7]
```diff
@@ -128,31 +130,43 @@ public void open(Map<String, Object> config, SourceContext sourceContext) throws
 
     @Override
     public void received(Consumer<T> consumer, Message<T> message) {
```
Comment (eolivelli) [inline pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/source/PulsarSource.java:132]: why do we need to change PulsarSource ?

**Review by eolivelli — CHANGES_REQUESTED** [review]
are you sure that you are not adding another patch on top of this patch ?
I see changed related to Pulsar IO and also to PulsarProducer that look unrelated

File: pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java (hunk 4 above)
Comment (congbobo184) [reply pulsar-functions/instance/src/main/java/org/apache/pulsar/functions/sink/PulsarSink.java:89]: PulsarSink use MessageImpl.getSchema()

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java (hunk 5 above)
Comment (congbobo184) [reply pulsar-client/src/main/java/org/apache/pulsar/client/impl/ProducerImpl.java:428]: same as above.

File: pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java (hunk 6 above)
Comment (congbobo184) [reply pulsar-functions/api-java/src/main/java/org/apache/pulsar/functions/api/Record.java:51]: i will change it.

**Comment by congbobo184** [comment]
@eolivelli please review again thanks.

**Review by eolivelli — APPROVED** [review]
Great!
I am happy the getReaderSchema helped to fix this problem

File: pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java [hunk 8]
```diff
@@ -1691,12 +1692,12 @@ protected void completeOpBatchReceive(OpBatchReceive<T> op) {
             initDeadLetterProducerIfNeeded();
             List<MessageImpl<T>> finalDeadLetterMessages = deadLetterMessages;
             MessageIdImpl finalMessageId = messageId;
-            deadLetterProducer.thenAccept(producerDLQ -> {
+            deadLetterProducer.thenAcceptAsync(producerDLQ -> {
```
Comment (mattisonchao) [inline pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java:1695]: Why do we use a common pool to run the next logic?

**Pull request merged**